ocaml/xapi/vm_sysprep.ml

```diff
@@ -0,0 +1,267 @@
+  let host = Helpers.get_localhost ~__context in
+  let hostname = Db.Host.get_hostname ~__context ~self:host in
+  match SR.find_opt ~__context ~label:(SR.name hostname) with
+  | None ->
+      ()
+  | Some sr -> (
+      Db.SR.get_VDIs ~__context ~self:sr
+      |> List.iter @@ fun self ->
+         match Db.VDI.get_record ~__context ~self with
+         | API.{vDI_VBDs= []; _} ->
+             call ~__context @@ fun rpc session_id ->
+             Client.VDI.destroy ~rpc ~session_id ~self
+         | _ ->
+             ()
+    )
+
+(** create a name with a random infix. We need random names for
+    temporary directories to avoid collisions of concurrent API calls *)
+let temp_name prefix suffix =
+  let rnd = Random.State.bits prng land 0xFFFFFF in
+  Printf.sprintf "%s%06x%s" prefix rnd suffix
+
+let temp_dir = Filename.get_temp_dir_name ()
+
+(** [mkdtmp] creates a directory in [dir] and returns its path. If [dir]
+    does not yet exist it is created. It is a an error if [dir] exists
+    and is not a directory. *)
+let mkdtemp ?(dir = temp_dir) ?(perms = 0o700) prefix suffix =
+  ( match Sys.file_exists dir with
+  | true when not (Sys.is_directory dir) ->
+      internal_error "s: %s is not a directory" __FUNCTION__ dir
+  | true ->
+      ()
+  | false ->
+      Unixext.mkdir_rec dir perms
+  ) ;
+  let rec try_upto = function
+    | n when n < 0 ->
+        internal_error "%s: can't create directory %S" __FUNCTION__ dir
+    | n -> (
+        let path = Filename.concat dir (temp_name prefix suffix) in
+        try Sys.mkdir path perms ; path with Sys_error _ -> try_upto (n - 1)
+      )
+  in
+  try_upto 20
+
+(** Crteate a temporary directory, and pass its path to [f]. Once [f]
+    returns the directory is removed again *)
+let with_temp_dir ?(dir = temp_dir) ?(perms = 0o700) prefix suffix f =
+  let dir = mkdtemp ~dir ~perms prefix suffix in
+  finally (fun () -> f dir) (fun () -> Unixext.rm_rec dir)
+
+(** name of the ISO we will use for a VM; this is not a path *)
+let iso_basename ~vm_uuid =
+  let now = Ptime_clock.now () |> Ptime.to_rfc3339 in
+  Printf.sprintf "sysprep-%s-%s.iso" vm_uuid now
+
+(** Create an ISO in [SR.dir] with content [unattend]. [SR.dir] is
+    created if it not already exists. Returns the path of the ISO image *)
+let make_iso ~vm_uuid ~unattend =
+  try
+    let basename = iso_basename ~vm_uuid in
+    let iso = SR.dir // basename in
+    Xapi_stdext_unix.Unixext.mkdir_rec SR.dir 0o755 ;
+    with_temp_dir ~dir:"/var/tmp/xapi" "sysprep-" "-iso" (fun temp_dir ->
+        let path = temp_dir // "unattend.xml" in
+        Unixext.write_string_to_file path unattend ;
+        debug "%s: written to %s" __FUNCTION__ path ;
+        let args = ["-r"; "-J"; "-o"; iso; temp_dir] in
+        Forkhelpers.execute_command_get_output genisoimage args |> ignore ;
+        (iso, basename)
+    )
+  with e ->
+    let msg = Printexc.to_string e in
+    Helpers.internal_error "%s failed: %s" __FUNCTION__ msg
```

Comment:
Backtrace is lost. But that is really the fault of the `internal_error` functions which calls `raise` after a formatting function.

To fix that we need either a version of `internal_error` that takes a `Printexc.raw_backtrace` as parameter, or call `Backtrace.is_important e` as the first statement in the `with e`. I suggest doing the latter.